src/test/framework/test_env_builder_kubernetes.js

```diff
@@ -1,61 +1,210 @@
+        const agents_yaml_path = path.join(output_dir, 'agents.yaml');
+        await fs.writeFileSync(agents_yaml_path, agents_yaml);
+        await kf.deploy_agents({
+            image: agent_image,
+            num_agents,
+            agents_yaml: agents_yaml_path,
+            envs: get_envs(),
+            cpu: agent_cpu,
+            mem: agent_mem,
+            pv
+        });
+
+        console.log(`waiting for ${num_agents} agents to be in optimal state`);
+        await wait_for_agents_optimal(mgmt_address, mgmt_ports['mgmt-https'], num_agents);
+        console.log(`all agents are in optimal state`);
+
+        // run test\remote command
+        let exit_code = 0;
         try {
-            await kf.delete_namespace({ namespace });
-            process.exit(0);
+            if (test) {
+                await run_test(test, server_details.services);
+            } else if (exec) {
+                console.log(`executing commad on server pod: ${exec}`);
+                await kf.kubectl(`exec ${server_details.pod_name} -- ${exec}`);
+            }
         } catch (err) {
-            console.error('got error on cleanup (clean only):', err);
-            process.exit(1);
+            if (err.code && Number.isInteger(err.code)) {
+                exit_code = err.code;
+            } else {
+                exit_code = 1;
+            }
+        }
+
+        if (clean && namespace_prefix) {
+            await kf.delete_namespace();
+        }
+
+        if (exit_code === 0) {
+            console.log('Completed successfully');
+        } else {
+            console.log('Completed with error');
         }
+        process.exit(exit_code);
+    } catch (err) {
+        exit_with_error('test_env_builder failed with error:', err);
     }
-    // await kf.create_server({ yaml_path: noobaa_core_yaml, namespace });
-    if (skip_agent_creation) {
-        console.log('skipping agents creation');
-    } else {
-        await kf.create_agent({ namespace });
+
+}
+
+async function run_test(test_file, services) {
+    try {
+        console.log(`running test ${test_file}`);
+        //pass as args all test_env args with addition of services info 
+        const args = process.argv.concat([
+            '--mgmt_ip', services.mgmt.address,
+            '--mgmt_port', services.mgmt.ports.mgmt,
+            '--mgmt_port_https', services.mgmt.ports['mgmt-https'],
+            '--s3_ip', services.s3.address,
+            '--s3_port', services.s3.ports.s3,
+            '--s3_port_https', services.s3.ports['s3-https'],
+        ]);
+        await promise_utils.fork(test_file, args);
+    } catch (err) {
+        console.log(`Test ${test_file} failed with error:`, err);
+        throw err;
     }
+}
+
+
+async function wait_for_agents_optimal(server_ip, server_port, expected_num_optimal, timeout) {
+    // default timeout of 5 minutes
+    timeout = timeout || 5 * 60000;
+    await P.resolve()
```

Comment:
Should we use P. ?

Reply:
I agree we should avoid using bluebird. I prefer to leave it for now so we can move on with the test env